### DISCUSSION TITLE
[Proposal] EqualityComparer should be configurable for record type properties

### DISCUSSION THREAD

**Discussion by SilentViking** [dotnet/csharplang#8245] [Language Ideas]
Value based equality check in record types are very useful, however any time a reference type property is introduced, the `Equals` and `GetHashCode` methods on the record must be overridden, or an implementation of the reference type must be written with `Equals` and `GetHashCode` overrides.

For example, collections are a common issue.

```
record MyRecord(IEnumerable<string> SomeList, string OtherProperty);
```

If I want value based equality on `SomeList`, IEnumerable<string> would need to be wrapped, or MyRecord will need to implement `Equals` and `GetHashCode`, losing out on the benefit of compile time generated code which would handle N number of other properties do have desirable default equality comparison.

Proposal:
An attribute, which can be applied to a property, which specifies an IEqualityComparer<T> type with a parameterless constructor.

```
record MyRecord(
   [property: Comparer<MyIEnumerableComparer<string>>()]
   IEnumerable<string> SomeList, 
    string OtherProperty);

public class MyIEnumerableComparer<T> : IEqualityComparer<IEnumerable<T>>
{
 // ...
}
```

Then the compile time generated Equals method will create an instance of MyIEnumerableComparer to compare SomeList, and otherwise use EqualityComparer<T>.Default.

**Comment by HaloFour** [comment]
This could be implemented with a source generators that interprets custom attributes and emits `Equals` and `GetHashCode` methods for the record.
